src/Microsoft.ML.Data/Commands/CrossValidationCommand.cs

```diff
@@ -28,8 +28,8 @@ public sealed class Arguments : DataCommand.ArgumentsBase
             [Argument(ArgumentType.Multiple, HelpText = "Trainer to use", ShortName = "tr")]
             public SubComponent<ITrainer, SignatureTrainer> Trainer = new SubComponent<ITrainer, SignatureTrainer>("AveragedPerceptron");
 
-            [Argument(ArgumentType.Multiple, HelpText = "Scorer to use", NullName = "<Auto>", SortOrder = 101)]
-            public SubComponent<IDataScorerTransform, SignatureDataScorer> Scorer;
+            [Argument(ArgumentType.Multiple, HelpText = "Scorer to use", NullName = "<Auto>", SortOrder = 101, SignatureType = typeof(SignatureDataScorer))]
```

Comment:
>SignatureDataScorer [](start = 134, length = 19)

I wonder if the information contained in SignatureType is identical to the information contained in typeof(Scorer) ?
As in, aren't these the type parameters you want to invoke with?

Reply:
The parameter information definitely overlaps, but this isn't duplicate information.

For example, if there are two signature delegates:

```C#
public delegate void SignatureSweeper();
public delegate void SignatureTrainer();
```

We can't discern which one we are trying to load from the component catalog solely on signature (both these signature delegates don't take any more parameters other than the Environment).

Reply:
Oh, I see. So the signature is needed for subselecting a list.

---
In reply to: [210097300](https://github.com/dotnet/machinelearning/pull/671#discussion_r210097300) [](ancestors = 210097300)

Reply:
The signature is needed in order to find the component in the ComponentCatalog.  We can have multiple components with the same name, but are different based on their "signature type".  So the components are keyed off both a name and a signature type:

https://github.com/dotnet/machinelearning/blob/e77f24e68d92f033910a5d783576d1879763d979/src/Microsoft.ML.Core/ComponentModel/ComponentCatalog.cs#L35-L41

https://github.com/dotnet/machinelearning/blob/e77f24e68d92f033910a5d783576d1879763d979/src/Microsoft.ML.Core/ComponentModel/ComponentCatalog.cs#L787